sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,347 @@
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+    private static final Logger LOG = LoggerFactory.getLogger(Write.class);
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileSystems.matchNewResource(outputDir, true));
+    }
+
+    public Write to(ResourceId outputDir) {
+      return toResource(ValueProvider.StaticValueProvider.of(outputDir));
+    }
+
+    public Write toResource(ValueProvider<ResourceId> outputDir) {
+      return toBuilder().setOutputDir(outputDir).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      input.apply(
+          ParDo.of(
+              new DoFn<KV<String, byte[]>, Void>() {
+                ValueProvider<ResourceId> tmpDir;
+
+                @Setup
+                public void setup() {
+                  tmpDir = ValueProvider.NestedValueProvider.of(
+                      getOutputDir(),
+                      new TemporaryDirectoryBuilder()
+                  );
+                }
+
+                @ProcessElement
+                public void processElement(ProcessContext c) throws IOException {
+                  KV<String, byte[]> kv = c.element();
+
+                  String filename = kv.getKey();
+                  ResourceId tmpFile = tmpDir.get().resolve(filename, RESOLVE_FILE);
+
+                  byte[] bytes = kv.getValue();
+                  try (
+                      OutputStream outStream =
+                          Channels.newOutputStream(FileSystems.create(tmpFile, BINARY))
+                  ) {
+                    outStream.write(bytes);
+                    outStream.flush();
+                  } catch (IOException e) {
+                    LOG.error(
+                        "Failed to write to temporary file [{}] for [{}].",
+                        tmpFile,
+                        getOutputDir().get().resolve(filename, RESOLVE_FILE)
+                    );
+                    FileSystems.delete(
+                        Collections.singletonList(tmpFile),
+                        MoveOptions.StandardMoveOptions.IGNORE_MISSING_FILES
+                    );
+                    throw e;
+                  }
+                }
+
+                @Teardown
+                public void teardown() throws IOException {
+                  try {
+                    // TODO: Replace FileSystems.rename() with a copy and delete to eliminate
+                    // empty directory restriction on Write.
+                    FileSystems.rename(
```

Comment:
Actually this is problematic. `@Setup` is called once per DoFn instance per worker - it's not a "global" initialization thing, and there's no guarantee that `@Teardown` will be called at all. Proper global initialization and cleanup are difficult, especially if you're writing an unbounded collection and there is no "global", there is only "per-window". Handling this properly would require borrowing a lot of complexity from `FileBasedSink`.

Maybe it'd be easier to do without a temporary directory, and instead create odd-looking temporary _files_ in the output directory, named something like `.beam-tmp.$filename.$timestamp.$unique_id`? Then you don't need to have a global initialization/finalization step for the directory. The downside is that if someone lists all files in `$output_dir/*`, they'll get the temp files too.

I suggest the following. Add a parameter `withTempDirectory`, which by default equals the output directory, and document that if you don't set it then potentially-garbage temp files will be written to the output directory, and if you don't like this, then you need to set `withTempDirectory`.